run.go

```diff
@@ -8,31 +8,41 @@ import (
 // Start will launch the go routines in the background to perform Autopilot.
 // When the context passed in is cancelled or the Stop method is called
 // then these routines will exit.
-func (a *Autopilot) Start(ctx context.Context) {
+func (a *Autopilot) Start(ctx context.Context) error {
 	a.runLock.Lock()
 	defer a.runLock.Unlock()
 
 	// already running so there is nothing to do
-	if a.running {
-		return
+	if a.execution != nil && a.execution.status == Running {
+		return nil
 	}
 
 	ctx, shutdown := context.WithCancel(ctx)
-	a.shutdown = shutdown
 	a.startTime = a.time.Now()
-	a.done = make(chan struct{})
-
-	// While a go routine executed by a.run below will periodically
-	// update the state, we want to go ahead and force updating it now
-	// so that during a leadership transfer we don't report an empty
-	// autopilot state. We put a pretty small timeout on this though
-	// so as to prevent leader establishment from taking too long
-	updateCtx, updateCancel := context.WithTimeout(ctx, time.Second)
-	defer updateCancel()
-	a.updateState(updateCtx)
-
-	go a.run(ctx)
-	a.running = true
+
+	exec := &execInfo{
+		status:   Running,
+		shutdown: shutdown,
+		done:     make(chan struct{}),
+	}
+
+	if a.execution == nil || a.execution.status == NotRunning {
+		// While a go routine executed by a.run below will periodically
+		// update the state, we want to go ahead and force updating it now
+		// so that during a leadership transfer we don't report an empty
+		// autopilot state. We put a pretty small timeout on this though
+		// so as to prevent leader establishment from taking too long. This
+		// only is done if we are not running and not shutting down so
+		// as to prevent conflicts with any autopilot routine just finishing
+		// up such as when restarting autopilot.
+		updateCtx, updateCancel := context.WithTimeout(ctx, time.Second)
+		defer updateCancel()
+		a.updateState(updateCtx)
```

Comment:
I feel a bit wary of having this outside of the `execMutex` lock now, is it possible to force an initial update in `runStateUpdater` instead?

Reply:
It will already run there. This was to prevent temporarily having the autopilot state disappear during a leader transition such as during an auto upgrade. Also the checks above to ensure that autopilot is in a not running state (while holding the runLock) should be enough to ensure no conflicts.

Basically the `runLock` is used to protect against multiple concurrent `Start` invocations as well as access to the execution status and done chans. The `execMutex` is essentially a leader lock to ensure only a single execution of autopilot go routines can be functional at any given time while still allowing `Start` to spawn the new go routines if the old ones are shutting down. I can add some more comments (and maybe renamed some fields) to make this clearer.

Reply:
Right, but can't we also prevent temporarily having the autopilot state disappear in `runStateUpdater` by doing an initial update before the ticker starts? https://github.com/hashicorp/raft-autopilot/blob/4a9e1e5540f01d7390bb97e6650ea42b4ac9c99d/run.go#L122-L133

I suppose that would mean waiting for a couple of goroutines to spin up, instead of immediately on `Start()`.

In any case, this is just me nitpicking code location, I can see that the conditional above ensures that we prevent conflicts here. 👍  Feel free to ignore.

Reply:
Exactly, its the waiting for the go routine to start and get scheduled that is the issue. If done inline in Start then updating the state would be finished before `establishLeadership` in Consul finishes.

I went ahead and added gaining the lock in Start too for good measure. It should be unnecessary but it also shouldn't be much overhead to gain the lock once.